src/sm2/field.rs

```diff
@@ -326,82 +339,160 @@ fn raw_mul(a: &FieldElem, b: &FieldElem) -> [u32; 16] {
     let mut carry: u64 = 0;
     let mut ret: [u32; 16] = [0; 16];
 
-    let mut k = 0;
-    while k < 15 {
-        let index = 15 - k;
-        let mut i = 0;
-        while i < 8 {
-            if i > k {
+    let mut ret_idx = 0;
+    while ret_idx < 15 {
+        let index = 15 - ret_idx;
+        let mut a_idx = 0;
+        while a_idx < 8 {
+            if a_idx > ret_idx {
                 break;
             }
-            let j = k - i;
-            if j < 8 {
-                let (u, v) = u32_mul(a.value[7 - i], b.value[7 - j]);
-                local += v;
-                carry += u;
+            let b_idx = ret_idx - a_idx;
+            if b_idx < 8 {
+                let (hi, lo) = u32_mul(a.value[7 - a_idx], b.value[7 - b_idx]);
+                local += lo;
+                carry += hi;
             }
 
-            i = i + 1;
+            a_idx += 1;
         }
         carry += local >> 32;
-        local = local & 0xffffffff;
+        local &= 0xffff_ffff;
         ret[index] = local as u32;
         local = carry;
         carry = 0;
 
-        k = k + 1;
+        ret_idx += 1;
     }
     ret[0] = local as u32;
     ret
 }
 
+impl ::std::cmp::PartialEq for FieldElem {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        let lhs = self.value;
+        let rhs = other.value;
+        if lhs[0] != rhs[0] {
+            return false;
+        }
+        if lhs[1] != rhs[1] {
+            return false;
+        }
+        if lhs[2] != rhs[2] {
+            return false;
+        }
+        if lhs[3] != rhs[3] {
+            return false;
+        }
+        if lhs[4] != rhs[4] {
+            return false;
+        }
+        if lhs[5] != rhs[5] {
+            return false;
+        }
+        if lhs[6] != rhs[6] {
+            return false;
+        }
+        if lhs[7] != rhs[7] {
+            return false;
+        }
+        true
+    }
+}
+
+impl ::std::cmp::Eq for FieldElem {}
+
+impl ::std::cmp::PartialOrd for FieldElem {
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<::std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl ::std::cmp::Ord for FieldElem {
+    #[inline]
+    fn cmp(&self, other: &Self) -> ::std::cmp::Ordering {
+        let lhs = self.value;
+        let rhs = other.value;
+        if lhs[0] != rhs[0] {
+            return if lhs[0] > rhs[0] {
```

Comment:
```
return lhs[0].cmp(rhs[0])
```

Reply:
OK, I will update this.

https://github.com/cryptape/libsm/blob/3f62eb92758df72eb8f5deec71c9702b82d22eaf/src/sm2/field.rs#L381-L393